src/mpool.c

```diff
@@ -409,35 +411,35 @@ void INTERNAL qt_mpool_free(qt_mpool pool,
     tc    = qt_mpool_internal_getcache(pool);
     cache = tc->cache;
     cnt   = tc->count;
-    qthread_debug(MPOOL_DETAILS, "->cache:%p (bt:%p) cnt:%u\n", cache, cache ? cache->block_tail : NULL, (unsigned int)cnt);
+    qthread_debug(MPOOL_DETAILS, "->cache:%p (bt:%p) cnt:%u\n", cache, cache ? atomic_load_explicit(&cache->block_tail, memory_order_relaxed) : NULL, (unsigned int)cnt);
     if (cache) {
         assert(cnt != 0);
-        n->next       = cache;
-        n->block_tail = cache->block_tail; // cache is likely to be IN cache, so this won't be slow
+        atomic_store_explicit(&n->next, cache, memory_order_relaxed);
+        atomic_store_explicit(&n->block_tail, atomic_load_explicit(&cache->block_tail, memory_order_relaxed), memory_order_relaxed); // cache is likely to be IN cache, so this won't be slow
```

Comment:
I'm a little skeptical that this is enough synchronization here, but the main goal of this PR is to explicitly require the current behavior we're getting implicitly via the x86 memory model so that the behavior will be the same across architectures. More refactoring later on is probably a good idea for stuff like this.

Reply:
It should be ok as tc is pointing to TLS which is thread local.